File: CGStudio/PGCore/Math/Torus3d.cs
namespace PG.Core.Math
{
    public class Torus3d
    {
        public double BigRadius { get; set; }
        public double SmallRadius { get; set; }
        public Vector3d Center { get; set; }

        public Torus3d()
        {
            this.Center = new Vector3d(0, 0, 0);
            this.BigRadius = 1.0;
            this.SmallRadius = 0.5;
        }

        public Torus3d(Vector3d center, double R, double r)
        {
            this.Center = center;
            this.BigRadius = R;
            this.SmallRadius = r;
        }

        public Vector3d GetPosition(double u, double v)
        {
            var t = u * 2.0 * System.Math.PI;
            var p = v * 2.0 * System.Math.PI;
            var x = BigRadius * System.Math.Cos(t) + SmallRadius * System.Math.Cos(p) * System.Math.Cos(t);
            var y = BigRadius * System.Math.Sin(t) + SmallRadius * System.Math.Cos(p) * System.Math.Sin(t);
            var z = SmallRadius * System.Math.Sin(p);
            return Center + new Vector3d(x, y, z);
        }
    }
}

File: Crystal/Math/Torus3d.cpp
#include "Torus3d.h"

#include "Tolerance.h"

using namespace Crystal::Math;

Torus3d::Torus3d() :
	center(0, 0, 0),
	R(1.0),
	r(0.5)
{}

Torus3d::Torus3d(const Vector3dd& center, const double R, const double r) :
	center(center),
	R(R),
	r(r)
{}

Vector3dd Torus3d::getPosition(const double u, const double v) const
{
	const auto t = u * 2.0 * PI;
	const auto p = v * 2.0 * PI;
	const auto x = R * ::cos(t) + r * ::cos(p) * ::cos(t);
	const auto y = R * ::sin(t) + r * ::cos(p) * ::sin(t);
	const auto z = r * ::sin(p);
	return center + Vector3dd(x, y, z);
}

Vector3dd Torus3d::getNormal(const double u, const double v) const
{
	// todo.
	return Vector3dd();
}
